internal/cache/file/downloader/downloader_test.go

```diff
@@ -0,0 +1,224 @@
+	"github.com/googlecloudplatform/gcsfuse/tools/integration_tests/util/operations"
+	. "github.com/jacobsa/ogletest"
+	"golang.org/x/net/context"
+)
+
+var cacheLocation string = path.Join(os.Getenv("HOME"), "cache/location")
+
+const DefaultPerm = os.FileMode(0644)
+
+func TestDownloader(t *testing.T) { RunTests(t) }
+
+type downloaderTest struct {
+	job         *Job
+	bucket      gcs.Bucket
+	object      gcs.MinObject
+	cache       *lru.Cache
+	fakeStorage storage.FakeStorage
+	fileSpec    data.FileSpec
+	jm          *JobManager
+}
+
+func init() { RegisterTestSuite(&downloaderTest{}) }
+
+func (dt *downloaderTest) SetUp(*TestInfo) {
+	locker.EnableInvariantsCheck()
+	operations.RemoveDir(cacheLocation)
+
+	// Create bucket in fake storage.
+	dt.fakeStorage = storage.NewFakeStorage()
+	storageHandle := dt.fakeStorage.CreateStorageHandle()
+	dt.bucket = storageHandle.BucketHandle(storage.TestBucketName, "")
+
+	dt.initJobTest(DefaultObjectName, []byte("taco"), 200, CacheMaxSize)
+	dt.jm = NewJobManager(dt.cache, DefaultPerm, cacheLocation, DefaultSequentialReadSizeMb)
+}
+
+func (dt *downloaderTest) TearDown() {
+	dt.fakeStorage.ShutDown()
+	operations.RemoveDir(cacheLocation)
+}
+
+func (dt *downloaderTest) verifyJob(job *Job, object *gcs.MinObject, bucket gcs.Bucket, sequentialReadSizeMb int32) {
+	job.mu.Lock()
+	defer job.mu.Unlock()
+	ExpectEq(object.Generation, job.object.Generation)
+	ExpectEq(object.Name, job.object.Name)
+	ExpectEq(bucket.Name(), job.bucket.Name())
+	downloadPath := dt.jm.getDownloadPath(getObjectPath(bucket.Name(), object.Name))
+	ExpectEq(downloadPath, job.fileSpec.Path)
+	ExpectEq(sequentialReadSizeMb, job.sequentialReadSizeMb)
+}
+
+func (dt *downloaderTest) Test_getObjectPath() {
+	inputs := [][]string{{"", ""}, {"a", "b"}, {"a/b/", "/c/d"}, {"", "a"}, {"a", ""}}
+	expectedOutPuts := [5]string{"", "a/b", "a/b/c/d", "a", "a"}
+
+	results := [5]string{}
+	for i := 0; i < 5; i++ {
+		results[i] = getObjectPath(inputs[i][0], inputs[i][1])
+	}
+
+	ExpectTrue(reflect.DeepEqual(expectedOutPuts, results))
+}
+
+func (dt *downloaderTest) Test_getDownloadPath() {
+	inputs := []string{"/", "a/b", "a/b/c/d", "/a", "a/"}
+	dt.jm.mu.Lock()
+	expectedOutputs := [5]string{dt.jm.cacheLocation, dt.jm.cacheLocation + "/a/b",
+		dt.jm.cacheLocation + "/a/b/c/d", dt.jm.cacheLocation + "/a", dt.jm.cacheLocation + "/a"}
+	dt.jm.mu.Unlock()
+
+	results := [5]string{}
+	for i := 0; i < 5; i++ {
+		results[i] = dt.jm.getDownloadPath(inputs[i])
+	}
+
+	ExpectTrue(reflect.DeepEqual(expectedOutputs, results))
+}
+
+func (dt *downloaderTest) Test_GetJob_NotExisting() {
+
+	// call GetJob for job which doesn't exist.
+	job := dt.jm.GetJob(&dt.object, dt.bucket)
+
+	dt.jm.mu.Lock()
+	defer dt.jm.mu.Unlock()
+	dt.verifyJob(job, &dt.object, dt.bucket, dt.jm.sequentialReadSizeMb)
+}
+
+func (dt *downloaderTest) Test_GetJob_Existing() {
+	// first create new job
+	expectedJob := dt.jm.GetJob(&dt.object, dt.bucket)
+	dt.jm.mu.Lock()
```

Comment:
Just curious, why are we taking jobManager resource lock, before calling verifyJob? We are not doing anything related to job-manager resources?

Reply:
We are passing dt.jm.sequentialReadSizeMb